File: pc-client/tests/crypto.test.ts
import { CryptoManager } from "../electron/crypto";

describe("CryptoManager", () => {
  test("key exchange: two managers derive the same shared secret", () => {
    const alice = new CryptoManager();
    const bob = new CryptoManager();

    // Exchange public keys
    const alicePub = alice.getPublicKeyBase64();
    const bobPub = bob.getPublicKeyBase64();

    // Both compute shared secret from the other's public key
    alice.computeSharedSecret(bobPub);
    bob.computeSharedSecret(alicePub);

    // Both should be ready
    expect(alice.isReady()).toBe(true);
    expect(bob.isReady()).toBe(true);

    // Verify they derived the same key by roundtripping a message
    const testMessage = { type: "test", data: "hello" };
    const encrypted = alice.encrypt(testMessage);
    const decrypted = bob.decrypt(encrypted);

    expect(decrypted).toEqual(testMessage);
  });

  test("encrypt/decrypt roundtrip", () => {
    const alice = new CryptoManager();
    const bob = new CryptoManager();

    alice.computeSharedSecret(bob.getPublicKeyBase64());
    bob.computeSharedSecret(alice.getPublicKeyBase64());

    const original = { type: "text", content: "Hello, E2EE world!" };
    const encrypted = alice.encrypt(original);

    // Encrypted wrapper should have the expected fields
    expect(encrypted).toHaveProperty("nonce");
    expect(encrypted).toHaveProperty("payload");
    expect(encrypted).toHaveProperty("tag");
    expect(typeof encrypted.nonce).toBe("string");
    expect(typeof encrypted.payload).toBe("string");
    expect(typeof encrypted.tag).toBe("string");

    // Decrypt should recover the original message
    const decrypted = bob.decrypt(encrypted);
    expect(decrypted).toEqual(original);

    // Also test the reverse direction
    const encrypted2 = bob.encrypt(original);
    const decrypted2 = alice.decrypt(encrypted2);
    expect(decrypted2).toEqual(original);
  });

  test("decrypt with wrong key returns null", () => {
    const alice = new CryptoManager();
    const bob = new CryptoManager();
    const eve = new CryptoManager();

    // Alice and Bob do key exchange
    alice.computeSharedSecret(bob.getPublicKeyBase64());
    bob.computeSharedSecret(alice.getPublicKeyBase64());

    // Eve does NOT exchange with Alice — she has a different key
    eve.computeSharedSecret(bob.getPublicKeyBase64());

    const original = { type: "secret", content: "classified" };
    const encrypted = alice.encrypt(original);

    // Eve should fail to decrypt (wrong key)
    const result = eve.decrypt(encrypted);
    expect(result).toBeNull();
  });

  test("decrypt with tampered payload returns null", () => {
    const alice = new CryptoManager();
    const bob = new CryptoManager();

    alice.computeSharedSecret(bob.getPublicKeyBase64());
    bob.computeSharedSecret(alice.getPublicKeyBase64());

    const original = { type: "text", content: "important message" };
    const encrypted = alice.encrypt(original);

    // Tamper with the ciphertext payload
    const tamperedPayload = Buffer.from(encrypted.payload, "base64");
    tamperedPayload[0] ^= 0xff; // flip bits in first byte
    const tampered = {
      ...encrypted,
      payload: tamperedPayload.toString("base64"),
    };

    const result = bob.decrypt(tampered);
    expect(result).toBeNull();
  });

  test("decrypt with tampered tag returns null", () => {
    const alice = new CryptoManager();
    const bob = new CryptoManager();

    alice.computeSharedSecret(bob.getPublicKeyBase64());
    bob.computeSharedSecret(alice.getPublicKeyBase64());

    const original = { type: "text", content: "important message" };
    const encrypted = alice.encrypt(original);

    // Tamper with the auth tag
    const tamperedTag = Buffer.from(encrypted.tag, "base64");
    tamperedTag[0] ^= 0xff;
    const tampered = {
      ...encrypted,
      tag: tamperedTag.toString("base64"),
    };

    const result = bob.decrypt(tampered);
    expect(result).toBeNull();
  });

  test("decrypt before key exchange returns null", () => {
    const alice = new CryptoManager();
    expect(alice.isReady()).toBe(false);

    const result = alice.decrypt({
      nonce: Buffer.alloc(12).toString("base64"),
      payload: Buffer.alloc(16).toString("base64"),
      tag: Buffer.alloc(16).toString("base64"),
    });
    expect(result).toBeNull();
  });

  test("encrypt before key exchange throws", () => {
    const alice = new CryptoManager();
    expect(() => alice.encrypt({ type: "test" })).toThrow(
      "Key exchange not completed",
    );
  });

  test("sha256 produces consistent hashes", () => {
    const data = Buffer.from("hello world");
    const hash1 = CryptoManager.sha256(data);
    const hash2 = CryptoManager.sha256(data);
    expect(hash1).toBe(hash2);
    expect(hash1).toMatch(/^[0-9a-f]{64}$/);
  });
});

File: mobile_client/test/crypto_test.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:flutter_test/flutter_test.dart';
import 'package:fast_share_mobile/crypto_service.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  group('CryptoService', () {
    test('key exchange: two services derive the same shared key', () async {
      final alice = CryptoService();
      final bob = CryptoService();

      await alice.init();
      await bob.init();

      // Exchange public keys
      final alicePub = await alice.getPublicKeyBase64();
      final bobPub = await bob.getPublicKeyBase64();

      // Both compute shared secret from the other's public key
      await alice.computeSharedSecret(bobPub);
      await bob.computeSharedSecret(alicePub);

      // Both should be ready
      expect(alice.isReady, isTrue);
      expect(bob.isReady, isTrue);

      // Verify they derived the same key by roundtripping a message
      const testMessage = {'type': 'test', 'data': 'hello'};
      final encrypted = await alice.encrypt(testMessage);
      final decrypted = await bob.decrypt(encrypted);

      expect(decrypted, equals(testMessage));
    });

    test('encrypt/decrypt roundtrip', () async {
      final alice = CryptoService();
      final bob = CryptoService();

      await alice.init();
      await bob.init();

      await alice.computeSharedSecret(await bob.getPublicKeyBase64());
      await bob.computeSharedSecret(await alice.getPublicKeyBase64());

      const original = {'type': 'text', 'content': 'Hello, E2EE world!'};
      final encrypted = await alice.encrypt(original);

      // Encrypted wrapper should have the expected fields
      expect(encrypted, contains('nonce'));
      expect(encrypted, contains('payload'));
      expect(encrypted, contains('tag'));
      expect(encrypted['nonce'], isA<String>());
      expect(encrypted['payload'], isA<String>());
      expect(encrypted['tag'], isA<String>());

      // Decrypt should recover the original message
      final decrypted = await bob.decrypt(encrypted);
      expect(decrypted, equals(original));

      // Also test the reverse direction
      final encrypted2 = await bob.encrypt(original);
      final decrypted2 = await alice.decrypt(encrypted2);
      expect(decrypted2, equals(original));
    });

    test('decrypt with wrong key returns null', () async {
      final alice = CryptoService();
      final bob = CryptoService();
      final eve = CryptoService();

      await alice.init();
      await bob.init();
      await eve.init();

      // Alice and Bob do key exchange
      await alice.computeSharedSecret(await bob.getPublicKeyBase64());
      await bob.computeSharedSecret(await alice.getPublicKeyBase64());

      // Eve does NOT exchange with Alice — she pairs with Bob instead
      await eve.computeSharedSecret(await bob.getPublicKeyBase64());

      const original = {'type': 'secret', 'content': 'classified'};
      final encrypted = await alice.encrypt(original);

      // Eve should fail to decrypt (wrong key)
      final result = await eve.decrypt(encrypted);
      expect(result, isNull);
    });

    test('decrypt with tampered payload returns null', () async {
      final alice = CryptoService();
      final bob = CryptoService();

      await alice.init();
      await bob.init();

      await alice.computeSharedSecret(await bob.getPublicKeyBase64());
      await bob.computeSharedSecret(await alice.getPublicKeyBase64());

      const original = {'type': 'text', 'content': 'important message'};
      final encrypted = await alice.encrypt(original);

      // Tamper with the ciphertext payload
      final payloadBytes = base64Decode(encrypted['payload']!);
      final tampered = Uint8List.fromList(payloadBytes);
      tampered[0] ^= 0xff; // flip bits in first byte

      final tamperedWrapper = <String, String>{
        'nonce': encrypted['nonce']!,
        'payload': base64Encode(tampered),
        'tag': encrypted['tag']!,
      };

      final result = await bob.decrypt(tamperedWrapper);
      expect(result, isNull);
    });

    test('decrypt with tampered tag returns null', () async {
      final alice = CryptoService();
      final bob = CryptoService();

      await alice.init();
      await bob.init();

      await alice.computeSharedSecret(await bob.getPublicKeyBase64());
      await bob.computeSharedSecret(await alice.getPublicKeyBase64());

      const original = {'type': 'text', 'content': 'important message'};
      final encrypted = await alice.encrypt(original);

      // Tamper with the auth tag
      final tagBytes = base64Decode(encrypted['tag']!);
      final tamperedTag = Uint8List.fromList(tagBytes);
      tamperedTag[0] ^= 0xff;

      final tamperedWrapper = <String, String>{
        'nonce': encrypted['nonce']!,
        'payload': encrypted['payload']!,
        'tag': base64Encode(tamperedTag),
      };

      final result = await bob.decrypt(tamperedWrapper);
      expect(result, isNull);
    });

    test('decrypt before key exchange returns null', () async {
      final alice = CryptoService();
      await alice.init();

      expect(alice.isReady, isFalse);

      final result = await alice.decrypt({
        'nonce': base64Encode(Uint8List(12)),
        'payload': base64Encode(Uint8List(16)),
        'tag': base64Encode(Uint8List(16)),
      });
      expect(result, isNull);
    });

    test('encrypt before key exchange throws', () async {
      final alice = CryptoService();
      await alice.init();

      expect(
        () => alice.encrypt({'type': 'test'}),
        throwsStateError,
      );
    });

    test('sha256 produces consistent hashes', () async {
      final data = utf8.encode('hello world');
      final hash1 = await CryptoService.sha256(data);
      final hash2 = await CryptoService.sha256(data);

      expect(hash1, equals(hash2));
      expect(hash1, matches(RegExp(r'^[0-9a-f]{64}$')));
    });
  });
}
